### PR TITLE
Add VideoMAE

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds [VideoMAE](https://github.com/MCG-NJU/VideoMAE), which extends [ViTMAE](https://huggingface.co/docs/transformers/model_doc/vit_mae) to videos. 

The only difference between VideoMAE and ViT is that you need to replace `nn.Conv2d` by `nn.Conv3d` in the patch embedding class. 😂 

To do:

- [ ] Decide on a name for `VideoMAEFeatureExtractor` (should we keep it, or rename to `VideoMAEProcessor`, `VideoMAEPreprocessor`?)
- [ ] Decide on the input format for video models; currently I've chosen `pixel_values` of shape (batch_size, num_frames, num_channels, height, width). The original implementation uses (B, C, T, H, W)
- [ ] Doc examples + tests
- [x] Incorporate changes of #17731 
- [ ] Make VideoMAEFeatureExtractor robust with return_tensors="np" by default, better tests

### CODE REVIEW

File: README_ko.md
Comment (sgugger) [inline README_ko.md:324]: Needs updating. I've made a PR to fix the bug where it's not updated from the main README, so should just be a `make fix-copies` in a couple of days.

File: src/transformers/__init__.py
Comment (sgugger) [inline src/transformers/__init__.py:792]: The new auto class should also be imported in the main init (and documented).

File: src/transformers/models/videomae/convert_videomae_to_pytorch.py
Comment (sgugger) [inline src/transformers/models/videomae/convert_videomae_to_pytorch.py:129]: To clean up.

File: src/transformers/models/videomae/convert_videomae_to_pytorch.py
Comment (sgugger) [inline src/transformers/models/videomae/convert_videomae_to_pytorch.py:142]: To clean up.

File: src/transformers/models/videomae/feature_extraction_videomae.py
Comment (sgugger) [inline src/transformers/models/videomae/feature_extraction_videomae.py:93]: For tensors, we should implement something using PyTorch here, as iterating through the frames will be super slow.

File: src/transformers/models/videomae/modeling_videomae.py
Comment (sgugger) [inline src/transformers/models/videomae/modeling_videomae.py:171]: Make one line per assignment. It will take less space and be easier to read.

File: src/transformers/models/videomae/modeling_videomae.py
Comment (sgugger) [inline src/transformers/models/videomae/modeling_videomae.py:788]: Put the comment above so the code can fit in one line.

File: src/transformers/models/videomae/modeling_videomae.py
Comment (sgugger) [inline src/transformers/models/videomae/modeling_videomae.py:792]: Here too

File: src/transformers/models/videomae/modeling_videomae.py
Comment (sgugger) [inline src/transformers/models/videomae/modeling_videomae.py:821]: The comment with abbreviations and one-letter names does not help (here and below).

File: src/transformers/models/videomae/test.py
Comment (sgugger) [inline src/transformers/models/videomae/test.py:1]: This file should not be added to the PR.

File: src/transformers/models/videomae/test_model.py
Comment (sgugger) [inline src/transformers/models/videomae/test_model.py:1]: This one shouldn't as well.

**Review by sgugger — COMMENTED** [review]
Thanks for working on this model, left a few comments.

File: src/transformers/models/videomae/feature_extraction_videomae.py
Comment (NielsRogge) [reply src/transformers/models/videomae/feature_extraction_videomae.py:93]: The original implementation also iterates over the frames (they use cv2 instead of Pillow for each frame): https://github.com/MCG-NJU/VideoMAE/blob/main/functional.py

Comment (NielsRogge) [reply src/transformers/models/videomae/feature_extraction_videomae.py]: Or do you mean when you provide a single tensor of shape (B, T, C, H, W)? Cause currently the feature extractor only accepts lists of PIL images or tensors

Comment (sgugger) [reply src/transformers/models/videomae/feature_extraction_videomae.py]: Normalization would be way faster if done on a big tensor, if we have tensors here. Likewise it would be faster done once an a big NumPy array if we have a NumPy arrays.

If we have a list of PIL images, it's converted anyway.

Comment (NielsRogge) [reply src/transformers/models/videomae/feature_extraction_videomae.py]: Oh yes sorry you probably only meant normalization. 

Should we add a `normalize_video` method to image_utils.py, that accepts either a list of NumPy arrays or PIL images?

Comment (sgugger) [reply src/transformers/models/videomae/feature_extraction_videomae.py]: That works too.

**Comment by fcakyon** [comment]
@NielsRogge do you have any ETA on this feature? I am developing a video classification fine-tuning framework, would love to use this model if it gets merged into main!

File: src/transformers/models/videomae/modeling_videomae.py
Comment (amyeroberts) [inline src/transformers/models/videomae/modeling_videomae.py:755]: Definition missing here

**Review by amyeroberts — APPROVED** [review]
LGTM! 🎥  Just one small nit comment.

Only question I have is about the `videomae/test.py` and `videomae/test_model.py` files and if they're in the right place, as they look more like scripts

File: src/transformers/models/videomae/modeling_videomae.py
Comment (sgugger) [inline src/transformers/models/videomae/modeling_videomae.py:469]: Fits in one or three lines.

File: src/transformers/models/videomae/modeling_videomae.py
Comment (sgugger) [inline src/transformers/models/videomae/modeling_videomae.py:729]: Fit in one or three lines.

File: src/transformers/models/videomae/test.py
Comment (sgugger) [inline src/transformers/models/videomae/test.py:1]: Still should not be there.

File: src/transformers/models/videomae/test_model.py
Comment (sgugger) [reply src/transformers/models/videomae/test_model.py:1]: Still should not be there.

**Review by sgugger — APPROVED** [review]
Make sure to remove all test scripts before merging (they really shouldn't be added in the first place, please take more care of the files you add with `git`).

File: src/transformers/image_utils.py
Comment (NielsRogge) [inline src/transformers/image_utils.py:31]: @sgugger is this ok?

**Comment by LysandreJik** [comment]
There seems to remain an issue with the docs:

```
Traceback (most recent call last):
  File "/usr/local/bin/doc-builder", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/site-packages/doc_builder/commands/doc_builder_cli.py", line 47, in main
    args.func(args)
  File "/usr/local/lib/python3.8/site-packages/doc_builder/commands/build.py", line 96, in build_command
    build_doc(
  File "/usr/local/lib/python3.8/site-packages/doc_builder/build_doc.py", line 405, in build_doc
    sphinx_refs = check_toc_integrity(doc_folder, output_dir)
  File "/usr/local/lib/python3.8/site-packages/doc_builder/build_doc.py", line 460, in check_toc_integrity
    raise RuntimeError(
RuntimeError: The following files are not present in the table of contents:
- model_doc/videomae
Add them to ../transformers/docs/source/en/_toctree.yml.
```

**Comment by NielsRogge** [comment]
@LysandreJik yes I was aware of that, should be fixed now.

Don't merge already please, I'm transferring checkpoints and updating the conversion script.

**Pull request merged**
